src/Services/ProgressService.php

```diff
@@ -44,17 +48,61 @@ public function getByUser(User $user): Collection
             }
             /** @var Group $group */
             foreach ($group->courses->where('status', '=', CourseStatusEnum::PUBLISHED) as $course) {
-                if (!$progresses->contains(fn (CourseProgressCollection $collection) => $collection->getCourse()->getKey() === $course->getKey())) {
+                if (!$progresses->contains(fn(CourseProgressCollection $collection) => $collection->getCourse()->getKey() === $course->getKey())) {
                     $progresses->push(CourseProgressCollection::make($user, $course));
                 }
             }
         }
 
         return $progresses
-            ->sortByDesc(fn (CourseProgressCollection $collection) => $collection->getCourse()->pivot->created_at)
+            ->sortByDesc(fn(CourseProgressCollection $collection) => $collection->getCourse()->pivot->created_at)
             ->values();
     }
 
+    public function getByUserPaginated(User $user, ?OrderDto $orderDto = null, ?int $perPage = 20): LengthAwarePaginator
+    {
+        $userId = $user->getKey();
+        $progresses = new Collection();
+
+        $query = Course::query()
+            ->leftJoinSub('SELECT course_id, MAX(created_at) as user_pivot_created_at FROM course_user GROUP BY course_id', 'course_user', function ($join) {
+                $join->on('courses.id', '=', 'course_user.course_id');
+            })
+            ->leftJoinSub('SELECT course_id, MAX(created_at) as group_pivot_created_at FROM course_group GROUP BY course_id', 'course_group', function ($join) {
+                $join->on('courses.id', '=', 'course_group.course_id');
+            })
+            ->whereHas('users', function (Builder $query) use ($userId) {
+                $query->where('users.id', $userId);
+            })
+            ->orWhereHas('groups', function (Builder $query) use ($userId) {
+                $query->whereHas('users', function (Builder $query) use ($userId) {
+                    $query->where('users.id', $userId);
+                });
+            });
+
+        $order = $orderDto->getOrder() ?? 'desc';
+
+        if ($orderDto->getOrderBy() && $orderDto->getOrderBy() !== 'obtained') {
+            $query->orderBy($orderDto->getOrderBy(), $order);
+        } else {
+            $query->orderByRaw("LEAST(COALESCE(user_pivot_created_at, '2999-12-31'), COALESCE(group_pivot_created_at, '2999-12-31')) $order");
```

Comment:
when using Raw queries, any values entered by the user should be binded to avoid SQL injection
```suggestion
            $query->orderByRaw("LEAST(COALESCE(user_pivot_created_at, '2999-12-31'), COALESCE(group_pivot_created_at, '2999-12-31')) ?", [$order]);
```
I am not entirely convinced about inserting a specific date for sorting here. Imho it would be better to write a separate solution depending on the database. Postgres sorts null values in the appropriate order for this case. I think that in MySQL you could sort first by whether both of the `user_pivot_created_at` and `group_pivot_created_at` values are null, and then by the value from the field e.g. `ISNULL(COALESCE(user_pivot_created_at, group_pivot_created_at)), COALESCE(user_pivot_created_at, group_pivot_created_at) ASC`. Here is example how to check DB https://github.com/EscolaLMS/Webinar/blob/main/src/Repositories/Criteria/WebinarIncomingCriterion.php#L25C20-L25C20

Reply:
I think its not possible to bind sorting direction like that because they are not considered as data but as sql syntax. I have tried but then it returns syntax error, anyway order column and direction are validated in request class